## 01. Import Libraries

In [60]:
# import libraries
import pandas as pd
import numpy as np
import os

## 02. Import Data

In [61]:
# folder path
path = r'/Users/lavinia/Documents/04-2020 Instacart Basket Analysis'

In [62]:
# import products.csv
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [63]:
# import orders_wrangled.csv
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

## 03. Data Consistency Checks

## Products 

## a) Mixed-Type Data

In [64]:
# create a dataframe
df_test = pd.DataFrame()

In [65]:
# create a new column named mix, filling with numeric, string, and boolean values
df_test['mix']=['a', 'b', 1, True]

In [66]:
# checking whether a dataframe contains any mixed-type columns
for col in df_test.columns.tolist():
    weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis=1)
    if len(df_test[weird]) > 0:
        print(col)

mix


In [67]:
# fixing the mixed-type data issue by converting the numeric to string
df_test['mix']=df_test['mix'].astype('str')

In [68]:
# Rechecking for mixed-type data
for col in df_test.columns.tolist():
    weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis=1)
    if len(df_test[weird]) > 0:
        print(col)
    else:
        print('all consistent')

all consistent


## b) Missing Values

In [69]:
# Finding the missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [70]:
# Create a subset dataframe with missing values
df_nan = df_prods[df_prods['product_name'].isnull()==True]

In [71]:
# Create a subset dataframe without missing values
df_prods_clean = df_prods[df_prods['product_name'].isnull()==False]

## c) Duplicates


In [72]:
# Finding the duplicates 
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [73]:
# Dropping the duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

## Exercise 4.5: Orders

In [74]:
# Q2: perform descriptive statistics
df_ords.describe()

order_id       user_id  order_number  order_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06       3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01       2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01       2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00       0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00       1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01       3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01       5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02       6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

## a) Mixed-Type Data

In [75]:
# Q3: checking whether the dataframe contains any mixed-type columns
for col in df_ords.columns.tolist():
    weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis=1)
    if len(df_ords[weird]) > 0:
        print(col)
    else:
        print(col, 'consistent')

order_id consistent
user_id consistent
order_number consistent
order_day_of_week consistent
order_hour_of_day consistent
days_since_prior_order consistent


Q4: There is no mixed-type data found across all columns

## b) Missing Values

In [77]:
# Q5: Checking for missing values
df_ords.isnull().sum()

order_id                       0
user_id                        0
order_number                   0
order_day_of_week              0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [78]:
# create a subset of dataframe that contains the missing values
df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull()==True]

In [86]:
# viewing the subset
df_ords_nan

order_id  user_id  order_number  order_day_of_week  \
0         2539329        1             1                  2   
11        2168274        2             1                  2   
26        1374495        3             1                  1   
39        3343014        4             1                  6   
45        2717275        5             1                  3   
...           ...      ...           ...                ...   
3420930    969311   206205             1                  4   
3420934   3189322   206206             1                  3   
3421002   2166133   206207             1                  6   
3421019   2227043   206208             1                  1   
3421069   3154581   206209             1                  3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 6 columns]

In [88]:
# understanding the reason behind the missing data
df_ords[df_ords['user_id'] == 1]

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0    2539329        1             1                  2                  8   
1    2398795        1             2                  3                  7   
2     473747        1             3                  3                 12   
3    2254736        1             4                  4                  7   
4     431534        1             5                  4                 15   
5    3367565        1             6                  2                  7   
6     550135        1             7                  1                  9   
7    3108588        1             8                  1                 14   
8    2295261        1             9                  1                 16   
9    2550362        1            10                  4                  8   
10   1187899        1            11                  4                  8   

    days_since_prior_order  
0                      NaN  
1                     15.0  
2                     21.0  
3                     29.0  
4                     28.0  
5                     19.0  
6                     20.0  
7                     14.0  
8                      0.0  
9                     30.0  
10                    14.0

In [89]:
# understanding the reason behind the missing data
df_ords[df_ords['user_id'] == 2]

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
11   2168274        2             1                  2                 11   
12   1501582        2             2                  5                 10   
13   1901567        2             3                  1                 10   
14    738281        2             4                  2                 10   
15   1673511        2             5                  3                 11   
16   1199898        2             6                  2                  9   
17   3194192        2             7                  2                 12   
18    788338        2             8                  1                 15   
19   1718559        2             9                  2                  9   
20   1447487        2            10                  1                 11   
21   1402090        2            11                  1                 10   
22   3186735        2            12                  1                  9   
23   3268552        2            13                  4                 11   
24    839880        2            14                  3                 10   
25   1492625        2            15                  1                 11   

    days_since_prior_order  
11                     NaN  
12                    10.0  
13                     3.0  
14                     8.0  
15                     8.0  
16                    13.0  
17                    14.0  
18                    27.0  
19                     8.0  
20                     6.0  
21                    30.0  
22                    28.0  
23                    30.0  
24                    13.0  
25                    30.0

In [95]:
# checking the count of unique user_id
user_id =df_ords['user_id'].nunique()
print(user_id)

206209


In [ ]:
df_ords.shape
# The number of row is 3,421,083 whereas 6.02% are missing.

There are 206,209 missing values in the "days_since_prior_order" column. These missing values are actually the first purchase (order_number = 1) of all Instacart customers (count of unique user_id equal with count of missing values), so there is no recorded data.

In [ ]:
# Q6: Addressing the missing values using an appropriate method. 

I decided to leave the missing values as it can be filtered out the by using "order_number =1"

## c) Duplicates

In [96]:
# Q7: checking for duplicate values
df_ords_dups = df_ords[df_ords.duplicated()]

In [98]:
df_ords_dups.shape

(0, 6)

There are no duplicates found

## 04. Exporting Data

In [82]:
# export the no_dups product file
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'), index=False)

In [100]:
# export the orders file
df_ords.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_checked.csv'), index=False)